feat: open packer logs from display

dhruvmanila opened the pull request:
fixes: #409 

This is similar to the diff window.

We could also implement a command like `PackerLog` to open the log file.

Also, should we open it instead in a new tab as the lines could get pretty long?

dhruvmanila committed:
feat: open packer logs from display

This is similar to the diff window.

akinsho commented on lua/packer/log.lua:
```diff
@@ -150,4 +150,10 @@ end
+function log.open()
+  -- This should be consistent with the format mentioned in log.new above.
+  local logfile = string.format('%s/%s.log', vim.fn.stdpath('cache'), default_config.plugin)
+  vim.cmd(string.format("pedit %s", logfile))
```
I wonder if preview is a good default for the log. As you mentioned in the issue this closes it seems like it will be quite hard to scan, but I haven't tried this locally, so I could be wrong.

dhruvmanila replied:
Yes, it would be. My suggestion would be to open it in a new tab and when the user will quit the tab, they will be redirected to the packer window. There's no additional code to be written here as with `tabedit`, the file will be opened in a new tab next to the current one and when that is closed, the user will be redirected to the previous tab.

nanotee replied:
Tiny suggestion: `pedit`, `tabedit` and other similar commands support a `+cmd` argument that allows you to position the cursor at a specific location. `tabedit + foo` positions the cursor at the end of file `foo`, which can be pretty useful for logs

dhruvmanila replied:
TIL! I always wondered what `+cmd` do. Thanks for the suggestion! :)

dhruvmanila committed:
refactor: open logfile in a new tab with nowrap

Also, the cursor is positioned at the end of the file.

akinsho commented on lua/packer/display.lua:
```diff
@@ -101,6 +101,10 @@ local keymaps = {
+  },
+  open_log = {
+    rhs = '<cmd>lua require"packer.display".open_log()<cr>',
+    action = 'open packer logs'
```
I'd reword this to `open log file` or something, all the commands are in the context of packer so seems redundant for this one command to explicitly mention packer

dhruvmanila replied:
Yeah, makes sense. Thanks for the suggestion!

akinsho reviewed (commented):
Just tried this out, and it works really well 👍🏾, I'd be happy to merge this (granted @wbthomason doesn't come back with some feedback) I just had one nit about how the command worded

dhruvmanila committed:
chore: reword action value for `open_log`

dhruvmanila committed:
docs: update display window keybindings

dhruvmanila commented:
Hi, @akinsho I've made the necessary changes. I've also updated the docs with the new keybindings and also added the documentation for the diff window key. Let me know if there's any other changes to make :)

wbthomason commented on lua/packer/log.lua:
```diff
@@ -150,4 +150,10 @@ end
+function log.open()
+  -- This should be consistent with the format mentioned in log.new above.
+  local logfile = string.format('%s/%s.log', vim.fn.stdpath('cache'), default_config.plugin)
+  vim.cmd(string.format("tabedit + %s", logfile))
```
Should we make the command for opening the log configurable? Maybe this should be tied to `config.display.open_fn`/`config.display.open_cmd`?

akinsho replied:
One thought regarding this is that though this would allow floating windows to work it means that the user would have to have their packer update buffer etc. work the same as their log buffer so if you used a tab for logs you'd also get a tab for a status buffer but that isn't really blocking I think it means people will have to stick with splits for logs

akinsho replied:
I'm sure there will be people who disagree (somewhat invalidating this comment 😆 ) but I think opening logs doesn't really make sense inside of a floating window, or a small split.

I appreciate that we're very much in the middle of the floating window trend but surely for readability of log files there isn't that much value to having them open in a bunch of different ways.

Anyway since people are likely to ask for flexibility regardless 🤷🏾

dhruvmanila replied:
No disrespect to people's preferences, but I don't think that should be configurable. The options to open are:
- Split: With this we would have 3 active splits in the current tabpage namely the current buffer, packer window and the log file. This would make viewing the logs uncomfortable with all the scrolling required.
- Floating window: In cases where the user wants to go back to either the packer window or the working buffer, the floating window will have to be closed and reopened again to view the logs.
- Tab: With the log file opened in a new tab, the user has the flexibility to move back and forth and closing the log file is just a single `q` away. It would also be comfortable to view the logs.

Similar examples: `:checkhealth`, `:LspInfo`
Although, the latter opens in a floating window.

---
This is not related to this PR but it seems that floating windows cannot be split further. This means if the user configured packer to be opened in a floating window and if they're trying to access the diff of a commit, it will open in a split in the previous buffer and the floating window will be on top of it. Thus, to access the diff window the user will have to close the floating window, but that would lose access to other diffs which the user would like to view.

nanotee replied:
Alternatively, making it a user command (e.g. `:PackerLog`) would allow supporting modifiers. Split by default, `:vertical PackerLog` to split vertically, `:tab PackerLog` to open a tab (IIRC Neovim is supposed to get a `:float` modifier at some point as well).

It's less configuration for users and it mimics existing Ex commands

wbthomason replied:
Thanks for the discussion; these are reasonable points. I think I favor @nanotee's suggestion to allow supporting modifiers via the `PackerLog` command, *but* to have the keybinding in the results window bound to `:tab PackerLog`, to open in a tab if the results split/float is already open (incidentally, I'm unsure how opening a new tab works when you're in a floating window - I assume it's as normal, but there might be an undesired interaction there?)

dhruvmanila replied:
That seems like a good solution. So, correct me if I'm wrong, but:
- `PackerLog` should just open the log file in the _current_ buffer to make the modifiers to work correctly.
- The default behavior in the _packer_ window should be to open in a new tab.

wbthomason commented on lua/packer/display.lua:
```diff
@@ -708,4 +727,8 @@ end
 display.ask_user = a.wrap(prompt_user)
 
+display.open_log = function()
+  if display.status.disp then display.status.disp:open_log() end
```
Perhaps we should separate displaying the log from having an active display? e.g. by opening a new display if there isn't already one, and additionally creating a command to open the log?

dhruvmanila replied:
> We could also implement a command like `PackerLog` to open the log file.

Do you mean something like this? That would be a good idea.

> by opening a new display if there isn't already one

I don't quite understand what you mean by this. Are you saying we should open a packer display if there isn't one? What if we just provide a command which calls `log.open()`?

wbthomason replied:
Yes, sorry - I saw your `PackerLog` comment right after I finished my review.

> I don't quite understand what you mean by this. Are you saying we should open a packer display if there isn't one? What if we just provide a command which calls log.open()?

Unless I misread the code, it currently only opens the log display if there is already a `packer` results window open, correct? What I was proposing was that the command should *always* display the logs whether or not there's already a `packer` window.

dhruvmanila replied:
> Unless I misread the code, it currently only opens the log display if there is already a packer results window open, correct? 

Yes, only in the packer window.

> What I was proposing was that the command should always display the logs whether or not there's already a packer window.

Right, understood.

wbthomason commented on lua/packer/display.lua:
```diff
@@ -571,6 +576,20 @@ local display_mt = {
+  open_log = function(self)
+    if not self:valid_display() then return end
+    if next(self.items) == nil then
+      log.info('Operations are still running; plugin info is not ready yet')
```
Should showing the log be tied to plugin operations running?

dhruvmanila replied:
If the user has not configured the buffer to be reloaded when the file changes outside of the editor, the logs will be static. (`:h autoread`, `:h checktime`) So, I thought it best to make the command available once packer has completed all the jobs.

wbthomason reviewed (changes requested):
Thanks! This looks like a useful feature. I left a few comments - mostly, I think that we don't need to tie displaying the logs to having an active display window (i.e. for filing an issue, people may want to just run a command to open the logs), and I think we should make the display for the logs either configurable or, ideally, tied to the existing `open_fn`/`open_cmd` config variables.

dhruvmanila commented:
Drop the ball on this one, apologies for that. I'm not sure if I'll ever be able to pick it up incase it's needed but it seems that the repository has a notice that it's unmaintained. In that case, I'll close this PR.
